csharp/src/Apache.Arrow.Adbc.FlightSql/FlightSqlStatement.cs

```diff
@@ -0,0 +1,191 @@
+﻿/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System;
+using System.Threading.Tasks;
+using Apache.Arrow.Adbc.Core;
+using Apache.Arrow.Flight;
+using Grpc.Core;
+
+namespace Apache.Arrow.Adbc.FlightSql
+{
+    /// <summary>
+    /// A Flight SQL implementation of <see cref="AdbcStatement"/>.
+    /// </summary>
+    public class FlightSqlStatement : AdbcStatement
+    {
+        private FlightSqlConnection flightSqlConnection;
+        
+        public FlightSqlStatement(FlightSqlConnection flightSqlConnection)
+        {
+            this.flightSqlConnection = flightSqlConnection;
+        }
+
+        public override async ValueTask<QueryResult> ExecuteQueryAsync()
+        {
+            FlightInfo info = await GetInfo(this.SqlQuery, this.flightSqlConnection.Metadata);
+
+            return new QueryResult(info.TotalRecords, new FlightSqlResult(this.flightSqlConnection, info));
+        }
+
+        public override QueryResult ExecuteQuery()
+        {
+            return ExecuteQueryAsync().Result;
+        }
+
+        public override UpdateResult ExecuteUpdate()
+        {
+            throw new NotImplementedException();
+        }
+
+        public async ValueTask<FlightInfo> GetInfo(string query, Metadata headers)
+        {
+            FlightDescriptor commandDescripter = FlightDescriptor.CreateCommandDescriptor(query);
+
+            return await this.flightSqlConnection.FlightClient.GetInfo(commandDescripter, headers).ResponseAsync;
+        }
+
+        /// <summary>
+        /// Gets a value from the Arrow array at the specified index using the Arrow field for metadata.
+        /// </summary>
+        /// <param name="arrowArray"></param>
+        /// <param name="field"></param>
+        /// <param name="index"></param>
+        /// <returns></returns>
+        public override object GetValue(IArrowArray arrowArray, Field field, int index)

```

Comment:
Similar to the explanation for ConvertArrowType, it's here for each driver to interpret the value. Again, in Snowflake, if the type is an Int64Array, but the logical type is _TIME_, then some special calculations need to be performed to convert that long value to a DateTime value in C#. There are several others, but these types of nuances are why I left it up to the drivers to implement.

Reply:
I'm still skeptical that the way to deal with this is to provide accessors for each driver, rather than having the driver convert things to the proper Arrow type in the first place.

Reply:
Though it's also possible that the C♯ library lacks efficient/convenient ways to implement such conversions (as compared to Go or C++).

Again at least for Snowflake the ADBC Go driver already does these conversions.

Reply:
At least, this implementation could be used as the default implementation, and then Flight SQL shouldn't need to override anything and other drivers could use it as a starting point?